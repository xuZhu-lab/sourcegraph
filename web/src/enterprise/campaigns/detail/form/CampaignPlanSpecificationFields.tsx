import React, { useCallback, useEffect } from 'react'
import combyJsonSchema from '../../../../../../schema/campaign-types/comby.schema.json'
import credentialsJsonSchema from '../../../../../../schema/campaign-types/credentials.schema.json'
import { ThemeProps } from '../../../../../../shared/src/theme'
import { MonacoSettingsEditor } from '../../../../settings/MonacoSettingsEditor'
import { CampaignType } from '../backend.js'

export const MANUAL_CAMPAIGN_TYPE = 'manual' as const

/**
 * Data represented in {@link CampaignPlanSpecificationFields}.
 */
export interface CampaignPlanSpecificationFormData {
    /** The campaign plan specification type (e.g., "comby"). */
    type: CampaignType | typeof MANUAL_CAMPAIGN_TYPE

    /** The campaign plan specification arguments (as JSONC). */
    arguments: string
}

interface Props extends ThemeProps {
    value: CampaignPlanSpecificationFormData
    onChange: (newValue: CampaignPlanSpecificationFormData) => void

    readOnly?: boolean
    className?: string
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const jsonSchemaByType: { [K in CampaignType]: any } = {
    comby: combyJsonSchema,
    credentials: credentialsJsonSchema,
}

const typeLabels: Record<CampaignType | typeof MANUAL_CAMPAIGN_TYPE, string> = {
    [MANUAL_CAMPAIGN_TYPE]: 'Manual',
    comby: 'Comby search and replace',
    credentials: 'Find leaked credentials',
}

const defaultInputByType: { [K in CampaignType]: string } = {
    comby: `{
    "scopeQuery": "repo:github.com/foo/bar",
    "matchTemplate": "",
    "rewriteTemplate": ""
}`,
    credentials: `{
    "scopeQuery": "repo:github.com/foo/bar",
    "matchers": [{ "type": "npm" }]
}`,
}

export const DEFAULT_CAMPAIGN_PLAN_SPECIFICATION_FORM_DATA: CampaignPlanSpecificationFormData = {
    type: 'comby',
    arguments: defaultInputByType.comby,
}

/**
 * Fields for selecting the type and arguments for the campaign plan specification.
 */
export const CampaignPlanSpecificationFields: React.FunctionComponent<Props> = ({
    value,
    onChange,
    readOnly,
    className,
    isLightTheme,
}) => {
    const onTypeChange = useCallback(
        (type: CampaignType | typeof MANUAL_CAMPAIGN_TYPE): void => {
            onChange({ type, arguments: type === MANUAL_CAMPAIGN_TYPE ? '' : defaultInputByType[type] })
        },
        [onChange]
    )
    const onArgumentsChange = useCallback((arguments_: string): void => onChange({ ...value, arguments: arguments_ }), [
        onChange,
        value,
    ])

    return (
        <div className={className}>
            <div className="flex-grow-1 form-group mb-0">
                {!readOnly ? (
                    <>
                        <label htmlFor="campaign-plan-specification-fields__type" className="font-weight-bold d-block">
                            Plan
                        </label>
                        <select
                            id="campaign-plan-specification-fields__type"
                            className="form-control w-auto d-inline-block e2e-campaign-type"
                            placeholder="Select campaign type"
                            onChange={e => onTypeChange(e.currentTarget.value as CampaignType)}
                            value={value.type}
                        >
                            {(Object.keys(typeLabels) as CampaignType[]).map(typeName => (
                                <option value={typeName || ''} key={typeName}>
                                    {typeLabels[typeName]}
                                </option>
                            ))}
                        </select>
                        {value.type === 'comby' ? (
                            <small className="ml-2">
                                <a rel="noopener noreferrer" target="_blank" href="https://comby.dev/#match-syntax">
                                    Learn about comby syntax
                                </a>
                            </small>
                        ) : value.type === MANUAL_CAMPAIGN_TYPE ? (
                            <small className="ml-2 text-muted">
                                You can add existing changesets on your code host to the campaign after creating it
                            </small>
                        ) : (
                            undefined
                        )}
                    </>
                ) : (
                    <p className="mb-0">{typeLabels[value.type || '']}</p>
                )}
            </div>
            {value.type !== MANUAL_CAMPAIGN_TYPE && (
                <MonacoSettingsEditor
                    className="flex-grow-1 mt-2 e2e-campaign-arguments"
                    isLightTheme={isLightTheme}
                    value={value.arguments}
                    jsonSchema={value.type ? jsonSchemaByType[value.type] : undefined}
                    height={110}
                    onChange={onArgumentsChange}
                    readOnly={readOnly}
                />
            )}
        </div>
    )
}
