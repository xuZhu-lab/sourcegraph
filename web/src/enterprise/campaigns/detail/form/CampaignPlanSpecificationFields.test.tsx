import React from 'react'
import { CampaignPlanSpecificationFields, MANUAL_CAMPAIGN_TYPE } from './CampaignPlanSpecificationFields'
import { createRenderer } from 'react-test-renderer/shallow'

jest.mock('../../../../settings/MonacoSettingsEditor', () => ({ MonacoSettingsEditor: 'MonacoSettingsEditor' }))

const PROPS = {
    onChange: () => undefined,
    isLightTheme: true,
}

describe('CampaignPlanSpecificationFields', () => {
    describe('manual type', () => {
        test('editable', () =>
            expect(
                createRenderer().render(
                    <CampaignPlanSpecificationFields
                        {...PROPS}
                        value={{ type: MANUAL_CAMPAIGN_TYPE, arguments: '' }}
                        readOnly={false}
                    />
                )
            ).toMatchSnapshot())

        test('read-only', () =>
            expect(
                createRenderer().render(
                    <CampaignPlanSpecificationFields
                        {...PROPS}
                        value={{ type: MANUAL_CAMPAIGN_TYPE, arguments: '' }}
                        readOnly={true}
                    />
                )
            ).toMatchSnapshot())
    })

    describe('non-manual type', () => {
        test('editable', () =>
            expect(
                createRenderer().render(
                    <CampaignPlanSpecificationFields
                        {...PROPS}
                        value={{ type: 'comby', arguments: '' }}
                        readOnly={false}
                    />
                )
            ).toMatchSnapshot())

        test('read-only', () =>
            expect(
                createRenderer().render(
                    <CampaignPlanSpecificationFields
                        {...PROPS}
                        value={{ type: 'comby', arguments: '' }}
                        readOnly={true}
                    />
                )
            ).toMatchSnapshot())
    })
})
