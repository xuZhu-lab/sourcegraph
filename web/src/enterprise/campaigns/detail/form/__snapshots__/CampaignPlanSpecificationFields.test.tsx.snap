// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampaignPlanSpecificationFields has initial value and calls onChange 1`] = `
<div>
  <div
    className="flex-grow-1 form-group mb-0"
  >
    <label
      className="font-weight-bold d-block"
      htmlFor="campaign-plan-specification-fields__type"
    >
      Plan
    </label>
    <select
      className="form-control w-auto d-inline-block e2e-campaign-type"
      id="campaign-plan-specification-fields__type"
      onChange={[Function]}
      placeholder="Select campaign type"
      value="comby"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="comby"
      >
        Comby search and replace
      </option>
      <option
        value="credentials"
      >
        Find leaked credentials
      </option>
    </select>
    <small
      className="ml-2"
    >
      <a
        href="https://comby.dev/#match-syntax"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn about comby syntax
      </a>
    </small>
  </div>
  <MonacoSettingsEditor
    className="flex-grow-1 mt-2 e2e-campaign-arguments"
    height={110}
    isLightTheme={true}
    jsonSchema={
      Object {
        "$id": "comby-spec.json#",
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": false,
        "description": "Schema for comby options",
        "properties": Object {
          "matchTemplate": Object {
            "description": "See https://comby.dev/#match-syntax for syntax",
            "minLength": 1,
            "type": "string",
          },
          "rewriteTemplate": Object {
            "description": "See https://comby.dev/#match-syntax for syntax",
            "minLength": 1,
            "type": "string",
          },
          "scopeQuery": Object {
            "description": "Define a scope to narrow down repositories affected by this change. Only GitHub and Bitbucket Server are supported.",
            "minLength": 1,
            "type": "string",
          },
        },
        "required": Array [
          "scopeQuery",
          "matchTemplate",
          "rewriteTemplate",
        ],
        "type": "object",
      }
    }
    onChange={[Function]}
    value="{
    \\"scopeQuery\\": \\"repo:github.com/foo/bar\\",
    \\"matchTemplate\\": \\"\\",
    \\"rewriteTemplate\\": \\"\\"
}"
  />
</div>
`;

exports[`CampaignPlanSpecificationFields manual type editable 1`] = `
<div>
  <div
    className="flex-grow-1 form-group mb-0"
  >
    <React.Fragment>
      <label
        className="font-weight-bold d-block"
        htmlFor="campaign-plan-specification-fields__type"
      >
        Plan
      </label>
      <select
        className="form-control w-auto d-inline-block e2e-campaign-type"
        id="campaign-plan-specification-fields__type"
        onChange={[Function]}
        placeholder="Select campaign type"
        value="manual"
      >
        <option
          value="manual"
        >
          Manual
        </option>
        <option
          value="comby"
        >
          Comby search and replace
        </option>
        <option
          value="credentials"
        >
          Find leaked credentials
        </option>
      </select>
      <small
        className="ml-2 text-muted"
      >
        You can add existing changesets on your code host to the campaign after creating it
      </small>
    </React.Fragment>
  </div>
</div>
`;

exports[`CampaignPlanSpecificationFields manual type read-only 1`] = `
<div>
  <div
    className="flex-grow-1 form-group mb-0"
  >
    <p
      className="mb-0"
    >
      Manual
    </p>
  </div>
</div>
`;

exports[`CampaignPlanSpecificationFields non-manual type editable 1`] = `
<div>
  <div
    className="flex-grow-1 form-group mb-0"
  >
    <React.Fragment>
      <label
        className="font-weight-bold d-block"
        htmlFor="campaign-plan-specification-fields__type"
      >
        Plan
      </label>
      <select
        className="form-control w-auto d-inline-block e2e-campaign-type"
        id="campaign-plan-specification-fields__type"
        onChange={[Function]}
        placeholder="Select campaign type"
        value="comby"
      >
        <option
          value="manual"
        >
          Manual
        </option>
        <option
          value="comby"
        >
          Comby search and replace
        </option>
        <option
          value="credentials"
        >
          Find leaked credentials
        </option>
      </select>
      <small
        className="ml-2"
      >
        <a
          href="https://comby.dev/#match-syntax"
          rel="noopener noreferrer"
          target="_blank"
        >
          Learn about comby syntax
        </a>
      </small>
    </React.Fragment>
  </div>
  <MonacoSettingsEditor
    className="flex-grow-1 mt-2 e2e-campaign-arguments"
    height={110}
    isLightTheme={true}
    jsonSchema={
      Object {
        "$id": "comby-spec.json#",
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": false,
        "description": "Schema for comby options",
        "properties": Object {
          "matchTemplate": Object {
            "description": "See https://comby.dev/#match-syntax for syntax",
            "minLength": 1,
            "type": "string",
          },
          "rewriteTemplate": Object {
            "description": "See https://comby.dev/#match-syntax for syntax",
            "minLength": 1,
            "type": "string",
          },
          "scopeQuery": Object {
            "description": "Define a scope to narrow down repositories affected by this change. Only GitHub and Bitbucket Server are supported.",
            "minLength": 1,
            "type": "string",
          },
        },
        "required": Array [
          "scopeQuery",
          "matchTemplate",
          "rewriteTemplate",
        ],
        "type": "object",
      }
    }
    onChange={[Function]}
    readOnly={false}
    value=""
  />
</div>
`;

exports[`CampaignPlanSpecificationFields non-manual type read-only 1`] = `
<div>
  <div
    className="flex-grow-1 form-group mb-0"
  >
    <p
      className="mb-0"
    >
      Comby search and replace
    </p>
  </div>
  <MonacoSettingsEditor
    className="flex-grow-1 mt-2 e2e-campaign-arguments"
    height={110}
    isLightTheme={true}
    jsonSchema={
      Object {
        "$id": "comby-spec.json#",
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": false,
        "description": "Schema for comby options",
        "properties": Object {
          "matchTemplate": Object {
            "description": "See https://comby.dev/#match-syntax for syntax",
            "minLength": 1,
            "type": "string",
          },
          "rewriteTemplate": Object {
            "description": "See https://comby.dev/#match-syntax for syntax",
            "minLength": 1,
            "type": "string",
          },
          "scopeQuery": Object {
            "description": "Define a scope to narrow down repositories affected by this change. Only GitHub and Bitbucket Server are supported.",
            "minLength": 1,
            "type": "string",
          },
        },
        "required": Array [
          "scopeQuery",
          "matchTemplate",
          "rewriteTemplate",
        ],
        "type": "object",
      }
    }
    onChange={[Function]}
    readOnly={true}
    value=""
  />
</div>
`;
